joshua/joshua_agent.py

```diff
@@ -581,16 +581,39 @@ def _run_ensemble(
                 # output = output.decode('utf-8')
 
                 break
-            except subprocess.TimeoutExpired:
+            except subprocess.TimeoutExpired as timeout_ex:
+                # TimeoutExpired exception contains partial output
+                # captured before timeout. We must preserve this!
                 if self._cancelled():
                     log("<cancelled>")
                     retcode = -1
-                    output = b""
+                    # Preserve any stdout captured before cancellation
+                    output = timeout_ex.stdout if timeout_ex.stdout else b""
                     break
                 if timeout_time and time.time() > timeout_time:
                     log("<timed out>")
+                    process.kill()
```

Comment:
I don't have a strong opinion here, but should we first try to `terminate()` the process (sends SIGTERM) and then send the `kill()` (sends SIGKILL) if it doesn't terminate after a fixed amount of time? That should give the process some additional time to write the logs before shutting down.